### PR TITLE
Tobin gh423 memory tests

### PR DESCRIPTION
**Pull request by tobine**
More changes for #423 
Some more memory tests and removing another invalid check.
Also updating doc validation to make un-implemented tests that don't have "TODO" in the document an error.

### CODE REVIEW

**Comment by tobine** [comment]
Force pushed a couple more updates and not planning any further development on this branch.
Have all the cases from #423 covered or killed and made a few updates to doc validation.

File: tests/layer_validation_tests.cpp [hunk 1]
```diff
@@ -996,6 +996,119 @@ TEST_F(VkLayerTest, TestAliasedMemoryTracking) {
...
+    mem_alloc.allocationSize = mem_reqs.size;
+    pass = m_device->phy().set_memory_type(mem_reqs.memoryTypeBits, &mem_alloc,
+                                           VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    if (!pass) {
+        vkDestroyImage(m_device->device(), image, NULL);
+        return;
+    }
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT,
+                                         " is aliased with buffer 0x");
+    // VALDIATION FAILURE due to image mapping overlapping buffer mapping
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Typo here

File: tests/layer_validation_tests.cpp [hunk 2]
```diff
@@ -996,6 +996,119 @@ TEST_F(VkLayerTest, TestAliasedMemoryTracking) {
...
+    image_create_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+    image_create_info.flags = 0;
+
+    VkMemoryAllocateInfo mem_alloc = {};
+    mem_alloc.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+    mem_alloc.pNext = NULL;
+    mem_alloc.allocationSize = 0;
+    mem_alloc.memoryTypeIndex = 0;
+
+    /* Create a mappable image.  It will be the texture if linear images are ok
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: This comment looks like it came from somewhere else - not doing anything with staging images here?

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -996,6 +996,119 @@ TEST_F(VkLayerTest, TestAliasedMemoryTracking) {
...
+
+    // Now correctly bind image to second mem allocation before incorrectly
+    // aliasing buffer
+    err = vkAllocateMemory(m_device->device(), &alloc_info, NULL, &mem_img);
+    ASSERT_VK_SUCCESS(err);
+    err = vkBindImageMemory(m_device->device(), image, mem_img, 0);
+    ASSERT_VK_SUCCESS(err);
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT,
+                                         " is aliased with image 0x");
+    err = vkBindBufferMemory(m_device->device(), buffer, mem_img, 0);
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: This is invalid even without the aliasing rule -- buffer was already bound to backing memory on :1047.

File: tests/layer_validation_tests.cpp [hunk 4]
```diff
@@ -996,6 +996,119 @@ TEST_F(VkLayerTest, TestAliasedMemoryTracking) {
...
+                                           VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    if (!pass) {
+        vkDestroyBuffer(m_device->device(), buffer, NULL);
+        return;
+    }
+    err = vkAllocateMemory(m_device->device(), &alloc_info, NULL, &mem);
+    ASSERT_VK_SUCCESS(err);
+
+    uint8_t *pData;
+    err = vkMapMemory(m_device->device(), mem, 0, mem_reqs.size, 0,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Do you actually need to fill the memory for this test?

File: tests/layer_validation_tests.cpp [hunk 5]
```diff
@@ -996,6 +996,119 @@ TEST_F(VkLayerTest, TestAliasedMemoryTracking) {
...
+    image_create_info.usage = VK_IMAGE_USAGE_TRANSFER_SRC_BIT;
+    image_create_info.queueFamilyIndexCount = 0;
+    image_create_info.pQueueFamilyIndices = NULL;
+    image_create_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+    image_create_info.flags = 0;
+
+    VkMemoryAllocateInfo mem_alloc = {};
+    mem_alloc.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+    mem_alloc.pNext = NULL;
+    mem_alloc.allocationSize = 0;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Perhaps move this block below GetImageMemoryRequirements on :1081 so you can just fill this field correctly once?

File: tests/layer_validation_tests.cpp [hunk 6]
```diff
@@ -996,6 +996,119 @@ TEST_F(VkLayerTest, TestAliasedMemoryTracking) {
...
+
+    vkGetBufferMemoryRequirements(m_device->device(), buffer, &mem_reqs);
+    VkMemoryAllocateInfo alloc_info = {};
+    alloc_info.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+    alloc_info.pNext = NULL;
+    alloc_info.memoryTypeIndex = 0;
+
+    // Ensure memory is big enough for both bindings
+    alloc_info.allocationSize = 0x10000;
+    pass = m_device->phy().set_memory_type(mem_reqs.memoryTypeBits, &alloc_info,
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: Being fussy (and the tests are /full/ of problems like this--) should use the intersection of the memoryTypeBits for the buffer and the image. I don't believe there's any guarantee that the intersection is nonzero, in which case tests like this should skip.

That's a fix for another day, though.

File: tests/layer_validation_tests.cpp (hunk 1 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: fixed

File: tests/layer_validation_tests.cpp (hunk 2 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: killed

File: tests/layer_validation_tests.cpp (hunk 3 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: done

File: tests/layer_validation_tests.cpp (hunk 4 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: no, killed

File: tests/layer_validation_tests.cpp (hunk 5 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: done

File: tests/layer_validation_tests.cpp (hunk 6 above)
Comment (tobine) [inline tests/layer_validation_tests.cpp]: cleaned this up a bit

**Comment by tobine** [comment]
pushed an additional CL with a few clean-ups

**Comment by chrisforbes** [comment]
:+1:

**Comment by tobine** [comment]
cherry-picked and pushed to master

**Pull request closed without merging**